### ISSUE TITLE
RLM-based family coherence audit (advisory, split-only)

### ISSUE THREAD

**Issue by Alex-Hunterz** [bandr-ai/bandits#15]
Clustering produces families deterministically and then nothing checks whether the grouping was any good. A human can run `merge-families` / `split-family`, but nothing tells them *which* families are worth looking at. This proposes a model-assisted audit that reads a whole family and reports on its coherence, without ever changing how grouping is computed.

## Why a Recursive Language Model rather than a plain LLM call

A family can hold dozens of traces. Feeding all members to one LLM call either truncates or blows the context window, and chunking by hand loses cross-chunk comparisons.

[Recursive Language Models](https://www.primeintellect.ai/blog/rlm) (Prime Intellect's `RLMEnv`, in their `verifiers` framework) load the input into a persistent Python REPL as a variable the root model never sees directly. The root model writes code to slice it and calls `llm_batch` to spawn sub-LLMs over the pieces, then combines results. That is the right shape for "read all 40 members of this family and tell me which ones do not belong."

Relevant properties from the writeup:

- recursion depth is currently fixed at 1
- latency is consistently higher than a plain LLM call
- the scaffold reportedly underperforms without RL training on it
- determinism is not addressed, and structurally cannot be assumed — the root model writes different Python on each run

The last point is the reason for the hard constraint below.

## Hard constraint: advisory only

The audit **must not** feed back into clustering. Grouping stays reproducible without any model:

- must not change `similarity`, `neighbors`, or any clustering parameter
- must not alter `family_id`, `fingerprint()`, `descriptor` used for identity, or the fit/held-out split
- must not merge families automatically (see asymmetry below)
- output is an annotation on `TaskFamily` — a limitation, proposed sub-groups, outlier trace ids, and optionally a legible name

Re-running `mine` without the audit must produce byte-identical families. The audit is a second, optional pass.

## Fragment, do not merge

Splitting and merging carry very different risk:

- **Splitting a family that should not have been split** yields two coherent families. Each drafts a valid verifier. Cost is some redundancy and lower per-family coverage. Recoverable, nothing downstream is corrupted.
- **Merging two families that should not be merged** yields one family with incoherent evidence. `draft_verifiers` groups `final_state_field` values that disagree and keys the check to whichever value happened to be most common. The result is a wrong verifier that looks fine.

So the audit may propose **splits**, and must leave over-merging alone unless a human decides otherwise. This matches what already exists: `split_family` is deterministic (it splits back into exact-instruction groups) while `merge_families` requires an explicit human call and records `proposed_by="human"`, `review_status="merged"`.

## Per-family, not a global parameter

An earlier version of this idea was "if the audit says a cluster is bad, raise or lower `neighbors`." That does not work: `neighbors` is global and cluster quality is local.

```
family A (refund)     coherent    -> wants n unchanged
family B (login)      fragmented  -> wants n higher
family C (bug fixes)  over-merged -> wants n lower
```

One knob, three conflicting demands. No value of `n` satisfies all three, because different regions of the corpus have different density.

Running the audit **independently per family**, and letting each family propose its own fragmentation, avoids the global parameter entirely. Family A is left alone, family B is annotated, family C proposes a split. Nothing global is touched.

## Stopping criterion

This is the part that needs deciding before implementation, and the issue should not be closed without it.

A naive loop — `cluster -> audit says bad -> change something -> re-cluster -> audit again` — has no stopping criterion. The audit is not calibrated against anything, so "keep going until the model is happy" optimises for the model's agreement rather than for correct grouping. That is the same failure `assess_promotion` exists to prevent one layer down: a check measured against nothing has agreed with nothing.

Options to decide between:

1. **Single pass, no loop (recommended starting point).** Audit runs once per family after clustering. Output is advisory. A human resolves. There is no loop, therefore no stopping criterion needed. Cheapest, safest, and sufficient to learn whether the audit is useful at all.
2. **Bounded fragmentation, single level.** A family may be split at most once by an audit-proposed split; the resulting sub-families are not re-audited in the same run. Bounded by construction. Matches RLM's depth-1 limit.
3. **Fixed-point on a deterministic predicate.** Loop until no family proposes a split, with a hard iteration cap. Only defensible once the audit is calibrated (see #16) — otherwise it terminates on model whim.
4. **Benchmark-gated.** Loop while pair-level precision/recall against a labelled benchmark improves, stop when it does not. Requires the benchmark to exist first.

Recommend shipping (1), and treating (2) as the next step. Do not implement (3) or (4) until the auditor has been measured.

## Naming comes free

The same call that reads a family can return a human-readable descriptor. Today a family is `family-c4052db0a1a0` with descriptor `"refund order <id>"` — a masked token string, not something to put in a report. A generated name such as `"Refund an eligible order"` is strictly better for review output.

Same constraint applies: presentation only. The generated name must never feed `family_id`, `fingerprint()`, clustering, or splits. Keep the mechanical descriptor alongside it.

## Suggested output shape

Per family, from one audit pass:

```python
class FamilyAudit(Contract):
    family_id: str
    coherent: bool
    outlier_trace_ids: tuple[str, ...] = ()
    proposed_subgroups: tuple[tuple[str, ...], ...] = ()
    generated_name: str | None = None
    rationale: str
    model: str
    prompt_digest: str
```

Persisted as its own derived artifact with the task set as parent, so a task set is never rewritten by an audit. Model and prompt digest recorded for the same reason #11 wants clustering provenance recorded.

## Acceptance criteria

- [ ] `mine` output is byte-identical with and without the audit
- [ ] the audit never merges families
- [ ] the audit never changes `similarity`, `neighbors`, or any clustering parameter
- [ ] proposed splits surface as annotations pointing at the existing `split-family` command
- [ ] generated names are presentation-only and never feed `family_id` or `fingerprint()`
- [ ] audit results persist as a derived artifact recording model and prompt digest
- [ ] a stopping criterion is chosen from the options above and documented before any loop is implemented

## Dependencies

- **#2** — with clustering inert on real corpora (45 of 46 families singleton), there is nothing meaningful to audit. Embeddings must land first.
- **#16** — the audit is an uncalibrated judge until measured against labelled same-family pairs. Anything beyond advisory output should wait on that benchmark.

## Out of scope

- LLM-decided membership, Docent-style (one YES/NO call per item/cluster pair). Non-reproducible grouping means a re-run silently produces a different training set.
- Automatic merging.
- Auto-tuning clustering parameters.
- Using generated names in verifier drafting or export.

**Comment by Alex-Hunterz** [comment]
---

## Decisions taken before implementation

### Scaffold: `dspy.RLM`, not `verifiers.RLMEnv`

The issue above named Prime Intellect's `RLMEnv` (in their `verifiers`
framework). Three implementations of the RLM idea exist, and `verifiers` is the
wrong one for this use:

| | `dspy.RLM` | `rlms` (Zhang, the paper's author) | `verifiers.RLMEnv` (Prime Intellect) |
| --- | --- | --- | --- |
| Purpose | inference | inference | RL training environments |
| REPL sandbox | Deno + Pyodide (WASM) | `local`, in-process `exec` by default | prime-sandboxes |
| Output typing | Signatures give typed fields | raw string, hand-parsed | rollout scoring |
| Bounds | `max_iters`, `max_llm_calls`, `max_output_chars` | iterations / budget / timeout | n/a |
| Runtime deps | ~12 (behind an optional extra) | ~6 | ~28, incl. `datasets`, `numpy`, `gepa`, `pyzmq` |

`verifiers.RLMEnv` extends `StatefulToolEnv` and expects a rollout-and-score
loop; this audit needs one inference call per family, so that framework's whole
reason for existing is unused while its dependency weight is not.

Between the two inference libraries, `dspy.RLM` is chosen over `rlms` for two
reasons that matter here specifically:

- **Sandboxing.** `rlms` defaults to `environment="local"`, which executes
  model-written Python in the host process — the same process holding the trace
  corpus this code audits. DSPy runs REPL code in a WASM sandbox by default.
- **Typed output.** `FamilyAudit` has typed fields. A DSPy Signature declares
  them; the alternative is hand-written JSON instructions and a tolerant parser,
  which is the most fragile part of a model-assisted pass.

Recorded as a trade, not a free win: DSPy pulls `litellm`, `gepa`, `diskcache`
and `cloudpickle`, and its default sandbox needs a `deno` binary. Both stay
behind an optional `audit` extra, so the core install keeps its three runtime
dependencies and CI never installs either.

### Stopping criterion: option 1, single pass, no loop

Required by the issue before implementation. The audit runs **once** per family
after clustering, emits advisory output, and a human resolves it. There is no
loop, so no stopping criterion is needed. Option 2 (bounded single-level
fragmentation) is the next step; options 3 and 4 stay blocked on #16, since a
loop that terminates on an uncalibrated judge's agreement is measuring the
judge, not the grouping.

### Model

`deepseek-v4-flash` for both root and sub-LLM, matching `judge.py`'s default.
Overridable per run; the model is recorded on every audit artifact, so a change
is visible rather than silent.

### Singleton families are skipped

A one-member family has nothing to split and no internal disagreement to find.
Auditing it spends a call to learn nothing, so singletons are skipped and
reported as skipped rather than silently absent.

### The audit runs inside `mine` by default

Running it always is what makes it useful, but it must not touch the task set.
The audit writes a separate `family_audit` artifact parented to the task set,
and `mine`'s own output is unchanged, preserving the byte-identical criterion
above. `--no-audit` skips the pass, which is what CI and offline runs use — it
is the escape hatch, not the default.

### `coherent` and `over_merged` are reported side by side

`FamilyCoherence.over_merged` is geometric — embedding diameter against the link
threshold. `FamilyAudit.coherent` is semantic — a model read the instructions.
They measure different things, so the report prints both without reconciling
them. Agreement is a strong signal; disagreement is the case worth a reviewer's
attention, and collapsing the two into one verdict would discard it.
